test/Table_Tests/src/Common_Table_Operations/Join/Union_Spec.enso

```diff
@@ -338,3 +339,24 @@ spec setup =
             t4 = table_builder [["A", [1.5]]]
             e5 = t1.union t4 allow_type_widening=False on_problems=Problem_Behavior.Ignore
             e5.should_fail_with No_Output_Columns
+
+        t1 = table_builder [["X", [0, 1, 2]], ["Y", ['aa', 'bb', 'cc']]] . cast "X" (Value_Type.Integer Bits.Bits_16) . cast "Y" (Value_Type.Char size=2 variable_length=False)
+        t2 = table_builder [["X", [3, 4, 5]], ["Y", ['x', 'y', 'z']]] . cast "X" (Value_Type.Integer Bits.Bits_32) . cast "Y" (Value_Type.Char size=1 variable_length=False)
+        supports_complex_types = (t1.is_error || t2.is_error || Problems.get_attached_warnings t1 . not_empty).not
+        if supports_complex_types then
+            Test.specify "should find a common type (2)" <|
+                t12 = t1.union t2
+                Problems.assume_no_problems t12
+                t12.at "X" . value_type . should_equal (Value_Type.Integer Bits.Bits_32)
+                t12.at "Y" . value_type . should_equal (Value_Type.Char size=2 variable_length=True)
+
+                t12.at "X" . to_vector . should_equal [0, 1, 2, 3, 4, 5]
+                t12.at "Y" . to_vector . should_equal ['aa', 'bb', 'cc', 'x', 'y', 'z']
+
+            Test.specify "should fail to find a common type if widening is not allowed (2)" <|
+                # TODO No_Output_Columns should have Column_Type_Mismatch as its cause for better readability
+                t1.union t2 allow_type_widening=False . should_fail_with No_Output_Columns
```

Comment:
This is a useful idea for future improvement, but I again don't want to hold this PR since it is not really related.

The idea is that `No_Output_Columns` here is a bit surprising - I did specify columns - there is none due to another error (`Column_Type_Mismatch`) - that's why I think it's worth ensuring that `No_Output_Columns` gets a `cause` field so that a more informative message can be done.

@jdunkerley shall we add it? I imagine we can create a separate ticket for it. It should take about half day of work.

Reply:
Please add as another ticket but yes having a cause makes sense.